Publish the style in the same workspace used for the datastore & layer

landryb opened the pull request:
This is something i've been running with since 2.8, against geoserver 2.3, and tested still working here on geonetwork 3 against geoserver 2.5.4 - rationale is to publish the style in the same workspace as the datastore & layer are published - it helps keeping things organized, and in certain cases (catalog privacy settings) users dont have access to 'global' styles (or if you want to disable the global endpoint), so it's better if the styles are grouped with their layers.

Would love feedback from @fxprunayre wether this would work with mapserver-backed-through-rest publishing.

fxprunayre commented:
> Would love feedback from @fxprunayre wether this would work with mapserver-backed-through-rest publishing.

Probably not, see https://github.com/neogeo-technologies/mra/blob/master/src/server.py#L821 but could be added.

landryb commented:
There's not much activity in mra unfortunately..... how should we proceed for this pr then ? have a flag differentiating mapserver/geoserver backends, or try to push the style into the workspace first, and then if it errors out in the global styles ? How would mra react to a POST on /workspaces/<workspace>/styles/<style> ? 404 ?

fxprunayre commented:
> have a flag differentiating mapserver/geoserver backends

Maybe add a boolean option to publish styles in the workspace defined or in the global space. For MRA, global space will be the option to choose ?

landryb commented:
The problem with adding 'options' is that you need to handle database schema migrations and all this nightmare...

landryb commented:
Or then i'll take the opportunity to add another boolean option for a mapserver to 'discover' workspaces on the fly, which is needed by https://github.com/geonetwork/core-geonetwork/wiki/Dynamic-Geoserver-nodes-for-geopublication...

landryb commented:
@fxprunayre - added the boolean 'pushstyleinworkspace' option, propagated it properly everywhere, and tested both cases (global vs workspace) against geoserver 2.5 here. I had a 'fun' time properly getting the checkbox ticked in angular... but should be good to merge now.

landryb commented:
Oh and of course it defaults to no/false so should be backwards-compatible with existing setups/mapserver out of the box.

landryb commented:
And while i was at improving this area, i reworked the createStyle() callers and added the possibility to publish the style from an sld file if found in the GeoFile object. I Tried hard to ensure state was still consistent, so if the sld is invalid it fallbacks to the default style as it was done before.

landryb commented:
@fxprunayre - should be good to go now, please review :)

landryb commented:
ping @fxprunayre :)

fxprunayre commented:
It should go first in 3.0.x which is then merged in develop.

fxprunayre commented:
Done in #1339
